pre/traits/function_traits.hpp

```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright 2014, Joel FALCOU <joel.falcou@lri.fr>
+ *
+ * \brief Provides access to lambda arity, return type, argumets type and access as std::function type:
+ *
+ *
+ * #include <iostream>
+ * #include <typeinfo>
+ * #include <pre/traits/function_traits.hpp>
+ *
+ * int main() {
+ *
+ *     auto my_lambda = [](bool arg_0, int arg_1, double arg_2, std::string arg_3) {
+ *                             return int{0} ;
+ *                         };
+ *
+ *     typedef pre::function_traits<decltype(my_lambda)> my_lambda_types;
+ *
+ *     std::cout << "my_lambda number of arguments " << my_lambda_types::arity                        << std::endl;
+ *     std::cout << "my_lambda return type is "      << typeid(my_lambda_types::result_type).name()   << std::endl;
+ *     std::cout << "my_lambda argument 0 type is "  << typeid(my_lambda_types::arg<0>).name()        << std::endl;
+ *     std::cout << "my_lambda argument 1 type is "  << typeid(my_lambda_types::arg<1>).name()        << std::endl;
+ *     std::cout << "my_lambda argument 2 type is "  << typeid(my_lambda_types::arg<2>).name()        << std::endl;
+ *     std::cout << "my_lambda argument 3 type is "  << typeid(my_lambda_types::arg<3>).name()        << std::endl;
+ *     std::cout << "my_lambda function type is "    << typeid(my_lambda_types::function_type).name() << std::endl;
+ *
+ *     auto std_function = pre::to_std_function(my_lambda);
+ *
+ *     std::cout << "my_lambda called as std::function "  << std_function(true, int{42}, double{3.14}, std::string{"Hello World"}) << std::endl;
+ *
+ *     return 0;
+ * }
+ */
+#ifndef PRE_TRAITS_FUNCTION_TRAITS_HPP
+#define PRE_TRAITS_FUNCTION_TRAITS_HPP
+
+ 
+#include <functional>
+#include <tuple>
+
+
+namespace pre {
```

Comment:
We can still make a convenience header with a using declaration, but I prefer things to be separated, as this library is kind of temporary library which should never grow and always go smaller as we should try to provide things to boost or remove them as soon as it's available in boost.